File: hashset.go
/*

Leetcode https://leetcode.com/problems/design-hashset/

In HashSet no duplicate elements are allowed
Time Complexity would be O(1) for add, contains 
Time Complexity would be O(1) for remove

Using map which is a data strcuture provided by golang to implement HashSet
map does not allow duplicate elements

*/type MyHashSet struct {
    store map[int]bool
}


func Constructor() MyHashSet {
    temp := make(map[int]bool)
    return MyHashSet{store : temp}
}


func (this *MyHashSet) Add(key int)  {
    this.store[key]= true
    
}

func (this *MyHashSet) Remove(key int)  {
    delete(this.store, key)
}


func (this *MyHashSet) Contains(key int) bool {
    if _, ok:=this.store[key]; ok {
        return true
    }
    return false
}


/**
 * Your MyHashSet object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Add(key);
 * obj.Remove(key);
 * param_3 := obj.Contains(key);
 */

File: HashSet.cs
#TC: O(1)

public class MyHashSet {

    private bool[][] storage; 
    private int buckets;      
    private int bucketItems;  

    public MyHashSet()
    {
        buckets = 1000; 
        bucketItems = 1000; 
        storage = new bool[buckets][];
    }

    
    private int Bucket(int key)
    {
        return key % buckets;
    }

    private int BucketItem(int key)
    {
        return key / bucketItems;
    }

    public void Add(int key)
    {
        int bucket = Bucket(key);
        if (storage[bucket] == null)
        {
            storage[bucket] = (bucket == 0) ? new bool[bucketItems + 1] : new bool[bucketItems];
        }
        int bucketItem = BucketItem(key);
        storage[bucket][bucketItem] = true;
    }

    public void Remove(int key)
    {
        int bucket = Bucket(key);
        if (storage[bucket] == null) return;
        int bucketItem = BucketItem(key);
        storage[bucket][bucketItem] = false;
    }

    public bool Contains(int key)
    {
        int bucket = Bucket(key);
        if (storage[bucket] == null) return false; 
        int bucketItem = BucketItem(key);
        return storage[bucket][bucketItem];
    }
}

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet obj = new MyHashSet();
 * obj.Add(key);
 * obj.Remove(key);
 * bool param_3 = obj.Contains(key);
 */